### PR TITLE
[WIP] C++ API

### PR DESCRIPTION
**Pull request by goldsborough**
This PR will be the development ground of PyTorch's C++ API for a while. To get CI feedback, without pushing to master, I will be posting continuous updates into this PR, one isolated, squashed commit at a time, and request reviews of individual commits.

CC @ebetica @jgehring

### CODE REVIEW

**Comment by goldsborough** [comment]
To begin, this PR creates a prototypical C++ API for `torch::nn::Module`. Lots of things are not yet implemented, however I created a stub for an LSTM module that showcases the general look and feel of the API, and have documented the methods I think `torch::nn::Module` should have. This PR also sets up the general directory structure, which I have chosen as `torch/csrc/api` as a start. I have the build system updates in another PR that I will push here once ready.

**Comment by dzhulgakov** [comment]
This is cool! Just making sure you're aware of https://sourcegraph.com/github.com/pytorch/pytorch/-/blob/torch/csrc/jit/script/module.h - we need to make sure that there's a single nn.Module abstraction in C++ which works with python, script, tracing, etc

File: torch/csrc/api/include/torch/nn/module.h [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+class Module {
+ public:
+  /// Tells the base `Module` about the name of the submodule.
+  explicit Module(std::string name);
+  virtual ~Module();
+
+  /// Returns the name of the `Module`.
+  const std::string& name() const noexcept;
+
+  /// Performs a recursive clone of the entire module hierarchy. This is to
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/module.h]: does it clone parameters too?

File: torch/csrc/api/include/torch/nn/module.h [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  /// enable deep copying in a polymorphic setting, i.e. when all you have is a
+  /// `Module*` but want to clone the whole module hierarchy.
+  virtual std::unique_ptr<Module> clone();
+
+  /// Takes a list of input variables and computes a list of output variables.
+  virtual std::vector<Tensor> forward(const std::vector<Tensor>& inputs) = 0;
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/module.h]: We should also have `operator()` (not virtual), so the interface will match what we have in Python. Right now this requires people to access `.forward`, which is discouraged.

File: torch/csrc/api/include/torch/nn/rnn.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      : CloneableModule<LSTM>("LSTM"),
+        weights_(at::randn(
+            torch::CPU(at::kFloat),
+            {3 * state_size, input_features * state_size})),
+        bias_(at::randn(torch::CPU(at::kFloat), {3 * state_size})) {
+    register_parameters({{"weights", weights_}, {"bias", bias_}});
+  }
+
+  std::vector<torch::Tensor> forward(
+      const std::vector<torch::Tensor>& inputs) override {
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/rnn.h]: meh, is the plan that all user defined modules will need to take in vectors as inputs? That sounds really dangerous (e.g. you don't do any checking below and can easily get segfaults for a bad call), and annoying (you need to parse out the args). Maybe we should let people define `forward` in the way they want, and then have a variadic template for the call operator (and forward the args as they are provided)

File: torch/csrc/api/include/torch/tensor.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace at {
+struct Type;
+enum class Backend;
+enum class ScalarType;
+} // namespace at
+
+namespace torch {
+/// There will only be gradient recording tensors in the frontend API.
+using Tensor = autograd::Variable;
```
Comment (apaszke) [inline torch/csrc/api/include/torch/tensor.h]: I'm still a bit worried that people will start mixing up tensor types (`torch::Tensor` and `at::Tensor), especially that they are implicitly convertible

File: torch/csrc/api/src/module.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  AT_ERROR("clone() has no default implementation");
+}
+
+// Train/Eval mode
+void Module::train() {
+  is_training_ = true;
+}
+
+void Module::eval() {
+  is_training_ = false;
```
Comment (apaszke) [inline torch/csrc/api/src/module.cpp]: Is there really any reason to avoid putting such trivial methods in the header? They really should be possible to inline

File: torch/csrc/api/include/torch/nn/rnn.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <vector>
+
+namespace torch { namespace nn {
+
+// This is largely just a proof-of-concept of the interface.
+class LSTM : public torch::nn::CloneableModule<LSTM> {
+ public:
+  LSTM(long input_features, long state_size)
+      : CloneableModule<LSTM>("LSTM"),
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/rnn.h]: The fact that you need to do
```cpp
MyModule() : ClonableModule<MyModule>("MyModule")
```
is a bit annoying. The density of `MyModule` is really high.

Is `clone` really that important? The non-CRTPed version would look so much nicer

File: torch/csrc/api/include/torch/nn/module.h [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+struct Archive;
+struct BufferCursor;
+struct ModuleCursor;
+struct ParameterCursor;
+
+/// The base class for all torch modules.
+class Module {
+ public:
+  /// Tells the base `Module` about the name of the submodule.
+  explicit Module(std::string name);
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/module.h]: Do we ever expect `name` to change at run time? If I understand correctly this is a replacement for the class name in Python, so it shouldn't change (and we could take a `const char*`)

**Comment by apaszke** [comment]
Nice POC! I found a few worrying things and left comments

File: torch/csrc/api/include/torch/tensor.h (hunk 4 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/tensor.h]: This was just to present my idea that there will only be `torch::Tensor` in this API. Users should neither know about the concept of `at::Tensor` nor `autograd::Variable`. We will have to be careful when migrating existing users like StarCraft, but for new users there should only be `torch::Tensor` and nothing else (like in Python)

File: torch/csrc/api/include/torch/nn/rnn.h (hunk 3 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/rnn.h]: Template methods can't be virtual, so there will be no way of declaring a virtual `forward` in the base class, and therefore we also can't call it from `operator()` 😢 Unless we do something nasty, like making `nn::Module` a CRTP too, then downcasting `this` inside `operator()` ...

File: torch/csrc/api/include/torch/nn/module.h (hunk 1 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/module.h]: Yeah it should be like `copy.deepcopy(module)` in Python

Comment (apaszke) [reply torch/csrc/api/include/torch/nn/module.h]: Can you please mention that in the comment? It's quite important

File: torch/csrc/api/include/torch/tensor.h (hunk 4 above)
Comment (apaszke) [reply torch/csrc/api/include/torch/tensor.h]: That's impossible, unless `torch::` will become an exact mirror of `at::`. Even you used `at::tanh` in your `LSTM` implementation, but this means that you silently introduced an `at::Tensor` into the mix.

File: torch/csrc/api/include/torch/nn/rnn.h (hunk 3 above)
Comment (apaszke) [reply torch/csrc/api/include/torch/nn/rnn.h]: Eh maybe we should just make `Module` a CRTP and solve the cloning, and other issues once and for all? It doesn't seem all that bad 😕

File: torch/csrc/api/include/torch/nn/module.h (hunk 7 above)
Comment (jgehring) [reply torch/csrc/api/include/torch/nn/module.h]: Wouldn't it be more convenient to specify a name when adding a module to another one, i.e. similar to the Python API's `add_module()` function? As far as I can see hat's the only place where you actually need a unique name.

**Review by jgehring — COMMENTED** [review]
This is perhaps a naive question, but why provide a `clone()` method rather than a copy constructor? `clone()` requires CRTP if you want to return the correct type, and in the current implementation it dictates the storage for the resulting module (unique_ptr).

**Comment by goldsborough** [comment]
It's for when all you have is an abstract Module*. In that case you need to know the static type to invoke the copy constructor. @ebetica was showing me some neuroevolution stuff where you guys needed this

On Apr 9, 2018, at 08:23, Jonas Gehring <notifications@github.com<mailto:notifications@github.com>> wrote:


@jgehring commented on this pull request.

This is perhaps a naive question, but why provide a clone() method rather than a copy constructor? clone() requires CRTP if you want to retrieve the correct type, and in the current implementation it dictates the storage for the resulting model (unique_ptr).

File: torch/csrc/api/include/torch/tensor.h (hunk 4 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/tensor.h]: I was thinking of doing `using namespace torch { using namespace at; using Tensor = autograd::Variable; }`.

File: torch/csrc/api/include/torch/nn/rnn.h (hunk 3 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/rnn.h]: Then we can't store them untemplatized. Like, Sequential will have a `vector<Module>` inside so the `Module` class should be untemplatized unless we want to templatize everything

Comment (ebetica) [reply torch/csrc/api/include/torch/nn/rnn.h]: It's pretty impossible to get varadic functions in C++. Even if you do use hacks to get it to accept different function signatures, if you have two models that output the same type of output and has different input types, and they both subclass a common super class, the compiler will be confused on which one to call.

I feel that a good compromise in this case is to use variants. I suggest we shoot for C++17 to make the code simpler, using structured bindings:

```c++
std::vector<torch::Variant> forward(const std::vector<torch::Variant>& inputs) override {
auto [ x, y, z ] = this->decode<int, torch::Tensor, std::string>(inputs);
}
```

https://skebanga.github.io/structured-bindings/

**Comment by ebetica** [comment]
@jgehring see the virtual copy constructor paradigm:

https://www.geeksforgeeks.org/advanced-c-virtual-copy-constructor/

File: torch/csrc/api/include/torch/nn/rnn.h (hunk 3 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/rnn.h]: We need to compile with C++11 unfortunately :/ We would have to pull in an external variant class. Also structured bindings do not work with vector, they only work with tuples and POD structs

Comment (ebetica) [reply torch/csrc/api/include/torch/nn/rnn.h]: This will possibly simply a lot of code too, if we're able to implement a few things like:

- Variant automatically behave like Variables unless casted
- Variants can be one of:
  - std::vector<Variant>
  - Variable
  - Tensor
  - common c types
  - std::unordered_map<Variant, Variant>
  - std::string

This way, the signature of the forward function is just
```c++
torch::Variant forward(const torch::Variant& inputs);
```

Comment (ebetica) [reply torch/csrc/api/include/torch/nn/rnn.h]: @goldsborough That's fine I think, even std::tie looks okay, the only disadvantage is that the inputs to the forward function will have to specified as a comment or something, which is exactly how it works with Python anyway.

**Comment by jgehring** [comment]
Ah thanks, of course, I completely missed that. Indeed, we discussed CRTP quite a lot for @ebetica's autogradpp...

File: torch/csrc/api/include/torch/nn/module.h (hunk 7 above)
Comment (apaszke) [reply torch/csrc/api/include/torch/nn/module.h]: I think the class name might be useful for printing?

File: torch/csrc/api/src/module.cpp (hunk 5 above)
Comment (apaszke) [reply torch/csrc/api/src/module.cpp]: I would be surprised if methods that have a single assignment inside increased the compile time significantly. Ok, we can just put everything in `.cpp` for now. I guess it's not a performance critical part.

File: torch/csrc/api/include/torch/tensor.h (hunk 4 above)
Comment (apaszke) [reply torch/csrc/api/include/torch/tensor.h]: I see, that makes sense

File: torch/csrc/api/include/torch/nn/rnn.h (hunk 3 above)
Comment (apaszke) [reply torch/csrc/api/include/torch/nn/rnn.h]: I wouldn't say that's how it works with Python, because the interpreter is actually doing all the unpacking and size checking for you, and this is the only problem I see with this approach. Such things are notoriously error-prone in C++, and as @goldsborough noted we can't rely on structured binding.

File: torch/csrc/api/include/torch/nn/module.h (hunk 7 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/module.h]: Yes it was primarily thought for pretty-printing, this constructor only gets invoked from the derived class constructor, to provide it with a name

Comment (jgehring) [reply torch/csrc/api/include/torch/nn/module.h]: But it's also used for the maps in the module so you'll have to make sure it's unique, e.g. with the code right now you can't add two LSTMs to a module.

File: setup.py [hunk 8]
```diff
@@ -543,6 +543,7 @@ def run(self):
 include_dirs += [
     cwd,
     os.path.join(cwd, "torch", "csrc"),
+    os.path.join(cwd, "torch", "csrc", "api", "include"),
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply setup.py]: My idea was users will want to just point their compiler at `torch/csrc/api/include`, and then use e.g. `#include <torch/nn/module.h>`, instead of the full path.

**Comment by ezyang** [comment]
CC @smessmer who is our resident C++ expert

File: tools/cpp_build/libtorch/CMakeLists.txt [hunk 9]
```diff
@@ -10,6 +10,11 @@ if (VERBOSE)
   message(STATUS "ATEN_BUILD_PATH is ${ATEN_BUILD_PATH}")
 endif()
 
+set(CMAKE_CXX_FLAGS "--std=c++11 -Wall -Wextra -Wno-unused-parameter -Wno-unused-variable ${CMAKE_CXX_FLAGS}")
```
Comment (ezyang) [inline tools/cpp_build/libtorch/CMakeLists.txt]: you don't have to change this, but cmake YI; list(APPEND ...) is faster than set().

File: torch/csrc/jit/script/module.h [hunk 10]
```diff
@@ -245,28 +186,28 @@ struct Module : public std::enable_shared_from_this<Module> {
...
   }
-  at::optional<Method&> find_method(const std::string& name) {
-    if(auto pm = methods.find(name))
-      return at::optional<Method&>(**pm);
-    return at::nullopt;
+  Method* find_method(const std::string& name) {
+    if (auto method = methods.find(name)) {
+      return method->get();
+    }
+    return nullptr;
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply torch/csrc/jit/script/module.h]: Sooo the thing is, it's technically not valid to have an `optional<&T>` (it's usually not ever valid to store a reference in a container, and `optional` is a container). There is an explanation [here](https://stackoverflow.com/questions/26858034/stdoptional-specialization-for-reference-types), and it is also mentioned in [cppreference.com](http://en.cppreference.com/w/cpp/utility/optional):

> There are no optional references; a program is ill-formed if it instantiates an optional with a reference type. Alternatively, an optional of a std::reference_wrapper of type T may be used to hold a reference.

It worked here because we weren't using the `optional<T&>` in a way that breaks it, e.g.

```
int x = 5;
at::optional<int&> o;
o.emplace(x);
o = x; // should work according to API of std::optional, works for references, but compiler error!

e.cpp:90:5: error: use of deleted function ‘typename std::enable_if<(! std::is_same<typename std::decay<_Tp>::type, at::optional<T&> >::value), at::optional<T&>&>::type at::optional<T&>::operator=(U&&) [with U = int&; T = int; typename std::enable_if<(! std::is_same<typename std::decay<_Tp>::type, at::optional<T&> >::value), at::optional<T&>&>::type = at::optional<int&>&]’
   o = x;
     ^
```

The solutions are:
1. Continue illegal use of `at::optional` in this way, hoping we won't use it in a way that breaks
2. Use `at::optional<std::reference_wrapper<T>>`, but it makes accessing the value harder (`*o` doesn't give you a `T` directly ...)
3. Use the original type responsible for representing "a nullable reference to something": pointers, old school pointers

@zdevito (author of said code)

File: torch/csrc/api/include/torch/nn/sequential.h [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+ private:
+  template <typename HeadType, typename... Tail>
+  void append(HeadType&& head, Tail&&... tail) {
+    append(std::forward<HeadType>(head));
+    // Recursively calls this method, until the parameter pack only thas this
+    // entry left. Then calls `append()` a final time.
+    append(std::forward<Tail>(tail)...);
+  }
+
+  std::vector<std::unique_ptr<Module>> modules_;
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/sequential.h]: 1) Fixed.
2) This is an important detail: The base (`Module`) cannot own the submodules. As @colesbury explained to me, the way we do this in Python is by custom descriptors that change `self.x` into `self._parameters['x']`. This doesn't work in C++ of course, so we either force our users to write code like `modules_["submodule"](inputs)` to access modules, or do it this way (i.e. keep `shared_ptr`s in the base class, and allow storage in the subclass so you write `submodule(inputs)`)

**Comment by goldsborough** [comment]
Changes:
1. `OrderedDict` now stores `Item`s, to support iteration (otherwise there's no easy way to go from a value back to its key)
2. Submodules are stored as `shared_ptr`s instead of `Module*`.
3. Implementation of cursors, which allow hierarchical traversal of submodules, parameters and buffers. Essentially a wrapper over `std::vector<std::pair<std::string, Module>>` but with a nice API to be able to write e.g. `module.parameters().find("key")` or `module.modules.apply([](Module& module) { ... })`. -- @smessmer may want to have a look
4. Quick hacky implementation of `zero_grad()` based on the cursors API. Need to think more about the Variable API, and implement `toType_()`

File: tools/cpp_build/libtorch/CMakeLists.txt (hunk 9 above)
Comment (goldsborough) [reply tools/cpp_build/libtorch/CMakeLists.txt]: I don't think this works -- `CMAKE_CXX_FLAGS` needs to be a (string separated) string, not a list

Comment (ebetica) [reply tools/cpp_build/libtorch/CMakeLists.txt]: You should prefer TARGET_COMPILE_OPTIONS I think.

File: tools/cpp_build/libtorch/CMakeLists.txt [hunk 12]
```diff
@@ -186,18 +195,36 @@ INSTALL(TARGETS torch
...
   "${COMMON_INCLUDES}")
 
 INSTALL(TARGETS test_jit
   RUNTIME DESTINATION "${TORCH_INSTALL_BIN_DIR}"
   LIBRARY DESTINATION "${TORCH_INSTALL_LIB_DIR}"
   ARCHIVE DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+
+## API tests
+
+SET(TORCH_API_TEST_SRCS ${TORCH_SRC_DIR}/csrc/api/test/test.cpp)
```
Comment (apaszke) [inline tools/cpp_build/libtorch/CMakeLists.txt]: Nooo please no tests inline with the code. Let's keep everything in `test/cpp`. `test_jit.cpp` is only a legacy thing that we shouldn't follow anymore

File: tools/cpp_build/libtorch/CMakeLists.txt [hunk 13]
```diff
@@ -186,18 +195,36 @@ INSTALL(TARGETS torch
...
+
+ADD_EXECUTABLE(test_api ${TORCH_API_TEST_SRCS})
+
+TARGET_LINK_LIBRARIES(test_api torch)
+
+TARGET_INCLUDE_DIRECTORIES(test_api PUBLIC
+  "${TORCH_SRC_DIR}/../aten/src/ATen/utils/catch/single_include"
+  "${COMMON_INCLUDES}")
+
+INSTALL(TARGETS test_api
```
Comment (apaszke) [inline tools/cpp_build/libtorch/CMakeLists.txt]: I doubt anyone will want to install tests 😄

File: torch/csrc/api/include/torch/nn/cursor.h [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  T* find(const std::string& key) noexcept;
+
+  /// Returns true if an item with the given `key` exists.
+  bool contains(const std::string& key) noexcept;
+
+  /// Counts the number of items available.
+  size_t size() const noexcept;
+
+ protected:
+  std::vector<Item> items_;
```
Comment (apaszke) [inline torch/csrc/api/include/torch/nn/cursor.h]: Why do `Cursor`s keep vectors of items? I thought they would just traverse the tree

File: torch/csrc/api/src/nn/module.cpp [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+  // parameters().apply([](Tensor& tensor) { tensor.toBackend_(kCPU); });
+}
+
+void Module::cuda() {
+  // parameters().apply([](Tensor& tensor) { tensor.toBackend_(kCUDA); });
+}
+
+void Module::type(at::ScalarType new_type) {
+  // parameters().apply([=](Tensor& tensor) { tensor.toType_(new_type); });
+  // buffers().apply([=](Tensor& tensor) { tensor.toType_(new_type); });
```
Comment (apaszke) [inline torch/csrc/api/src/nn/module.cpp]: These are implemented now, right?

File: torch/csrc/jit/script/module.h [hunk 16]
```diff
@@ -1,8 +1,10 @@
 #pragma once
+
 #include "torch/csrc/jit/ir.h"
 #include "torch/csrc/jit/graph_executor.h"
 #include "torch/csrc/autograd/variable.h"
-#include <ATen/optional.h>
+
+#include "torch/csrc/api/include/torch/detail/ordered_dict.h"
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/script/module.h]: We can shorten this to `torch/detail/ordered_dict.h`, because you've added the include path, right?

File: torch/csrc/jit/script/module.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/script/module.h]: Ok, let's use the pointer...

File: torch/csrc/jit/script/init.cpp [hunk 17]
```diff
@@ -309,23 +309,20 @@ void initJitScriptBindings(PyObject* module) {
...
-          auto & p = parameters[i];
-          py::tuple r(3);
-          result[i] = std::make_tuple(
-            p.name,
-            static_cast<const autograd::Variable&>(*p.slot()),
-            p.is_buffer);
-
+        py::tuple result(self.get_parameters().size());
+        size_t index = 0;
+        for (auto& parameter : self.get_parameters()) {
```
Comment (apaszke) [inline torch/csrc/jit/script/init.cpp]: Can we please use regular `for` loops when we iterate over multiple things? I find the mutable `index` pattern really annoying, as it only works if you iterate over 2 things. I'd stick to a rule of 1 item = range-based, 2+ items = old-style for loop.

File: torch/csrc/jit/script/init.cpp [hunk 18]
```diff
@@ -348,9 +345,12 @@ void initJitScriptBindings(PyObject* module) {
...
       .def("_method_names", [](Module& self) {
-        return fmap(self.get_methods(), [](const std::unique_ptr<Method> & m) {
-          return m->name();
-        });
+        std::vector<std::string> method_names;
+        method_names.reserve(self.get_methods().size());
+        for (auto& method : self.get_methods()) {
+          method_names.push_back((*method)->name());
+        }
+        return method_names;
```
Comment (apaszke) [inline torch/csrc/jit/script/init.cpp]: What's wrong with `fmap`?

**Review by apaszke — COMMENTED** [review]
Mostly LGTM. Let's get this in soon, because it's getting dangerously large.

File: torch/csrc/api/include/torch/nn/cursor.h (hunk 14 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/cursor.h]: So I had this first, but it adds some complication. Primarily, doing iterative DFS requires parent pointers, and that makes it super hard/dangerous to do things like moves (since all the parent pointers have to be updated). I discussed this with Zach, and we'll never really have more than, say, 100 elements in the hierarchy right (and maybe 10 in practice)? So highly likely this will be fast enough for anyone (and the code is so much simpler). So I figured we'll start with the simple solution, and can always add back the more complex one if it's unbearable

File: torch/csrc/api/src/nn/module.cpp (hunk 15 above)
Comment (goldsborough) [reply torch/csrc/api/src/nn/module.cpp]: I'm waiting for the "Autograd API" in Tensor. I need the `set_type` behavior, which is currently `temporary_hack_set_type` and will be `detail::set_data` in the autograd API PR

File: torch/csrc/jit/script/init.cpp (hunk 17 above)
Comment (goldsborough) [reply torch/csrc/jit/script/init.cpp]: ok sounds good

File: torch/csrc/jit/script/init.cpp (hunk 18 above)
Comment (goldsborough) [reply torch/csrc/jit/script/init.cpp]: Can do. When we replace the script module with the new module, we'll be able to do `module.methods().apply()` hopefully

File: torch/csrc/api/include/torch/nn/cursor.h (hunk 14 above)
Comment (apaszke) [reply torch/csrc/api/include/torch/nn/cursor.h]: Right, let's keep it.

File: tools/cpp_build/libtorch/CMakeLists.txt (hunk 12 above)
Comment (goldsborough) [reply tools/cpp_build/libtorch/CMakeLists.txt]: Oh ok, sure. I was unsure where to put it. I will create a `tests/cpp` folder 👍

File: torch/csrc/api/src/nn/module.cpp (hunk 15 above)
Comment (apaszke) [reply torch/csrc/api/src/nn/module.cpp]: Cool, thanks for the explanation!

**Comment by goldsborough** [comment]
@apaszke I was actually planning on doing all development in this PR, and squashing commits to make it manageable. This means we don't have to make it "public"  yet (so people don't use it). And I think if I push it onto a branch, say `pytorch:cpp` , the CI won't test it when I make PRs (is this the case?)? Is this a good idea?

**Comment by apaszke** [comment]
I think it's easier to merge this thing in parts, even if it's going to undergo breaking changes. It's already 1,1k lines long, and reviewing PRs that have 2k+ is really hard.

It's probably ok to merge this even before it's public. I doubt it's going to be used before we actually start advertising it, and we should make it clear that it's really experimental and will break your code.

**Review by apaszke — CHANGES_REQUESTED** [review]
Just to make sure this isn't merged before you remove the accidental ONNX submodule change.

File: torch/csrc/api/include/torch/nn/rnn.h (hunk 3 above)
Comment (dzhulgakov) [reply torch/csrc/api/include/torch/nn/rnn.h]: Is it possible to just have 2 versions of forward()? one receiving variants and one receiving list of tensors? Then the variant one would do down-casting and checking.

File: torch/csrc/api/include/torch/tensor.h (hunk 4 above)
Comment (dzhulgakov) [reply torch/csrc/api/include/torch/tensor.h]: Can it be reconciled with the API that jit right now takes (seemingly enforcing everything to be variables too): https://sourcegraph.com/github.com/pytorch/pytorch/-/blob/torch/csrc/jit/variable_tensor_list.h#L8:8

**Comment by goldsborough** [comment]
Will be rebased on top of autogradpp
